### PR TITLE
KAFKA-5011: down convert the message on replication if necessary

### PR DESCRIPTION
**Pull request by becketqin**

### CODE REVIEW

File: core/src/main/scala/kafka/log/Log.scala [hunk 1]
```diff
@@ -513,6 +513,11 @@ class Log(@volatile var dir: File,
           // we are taking the offsets we are given
           if (!appendInfo.offsetsMonotonic || appendInfo.firstOffset < nextOffsetMetadata.messageOffset)
             throw new IllegalArgumentException("Out of order offsets found in " + records.records.asScala.map(_.offset))
+
+          if (!records.hasCompatibleMagic(config.messageFormatVersion.messageFormatVersion)) {
+            trace(s"Down converting messages on replica fetching for $topicPartition")
+            validRecords = records.downConvert(config.messageFormatVersion.messageFormatVersion)
+          }
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/Log.scala:520]: Does it make sense to update the `leaderEpochCache` and then down-convert? cc @junrao

File: core/src/test/scala/unit/kafka/log/LogTest.scala [hunk 2]
```diff
@@ -1758,6 +1758,19 @@ class LogTest {
     recoveredLog.close()
   }
 
+  @Test
+  def testDownConversionForReplication(): Unit = {
+    val props = new Properties()
+    props.setProperty(LogConfig.MessageFormatVersionProp, KAFKA_0_10_0_IV1.version)
+    val log = new Log(logDir, LogConfig(props), recoveryPoint = 0L, scheduler = time.scheduler, time = time)
+    val recordBachV2 = singletonRecordsWithLeaderEpoch(value = "random".getBytes, leaderEpoch = 1, offset = 0)
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/log/LogTest.scala:1766]: recordBatchV2 (typo)

**Review by ijuma — COMMENTED** [review]
Thanks for the PR, left a couple of minor comments. Would be good for @junrao to check that this is OK.

File: core/src/main/scala/kafka/log/Log.scala (hunk 1 above)
Comment (becketqin) [reply core/src/main/scala/kafka/log/Log.scala:520]: It seems `leaderEpochCache` has already been updated in line 508? Am I missing something. I was trying to keep the metadata update unchanged, just down convert the message before appending to the log.

**Pull request closed without merging**